README
Deco version 0.10
=================

The Deco package can be used to simulate tissue behaviour during scuba diving.

This is very much work in progress. Results are by far not stable yet.

WARNING !!!

Do not base your actual dives on the values obtained from this software. It has been written for educational and fun purposes only and has NOT BEEN BACKED by any dives made.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

   Test::More
   Test::Exception
   GD::Graph
   Config::General

You can force install the package without the Test::* modules available. The GD::Graph module is only needed when you want to use the Dive::plot package.

COPYRIGHT AND LICENCE

Copyright (C) 2006 by Jaap Voets

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Deco.

0.11  2006-10-16
      Added percentage() function to the Deco::Dive::Plot package to print out percentage saturation graph
      Started on finding time limits for full table with repetitive dive groups

0.10  2006-10-12
      Started on the Deco::Dive::Table package to produce no decompression limit tables for given depths. 
      Added the padi config.
      Documented all public functions, so passing pod-coverage test.

0.09  2006-08-15 
      Added point function to Deco::Dive to set depth points manually. Got the buhlmann config included
      Tried testing against published buhlmann no stop limits, but results still differ.

0.08  2006-07-10 
      Added suggested links to POD.

0.07  2006-07-07
      Added gas functions to tissue and dive, to set the gas mix manually

0.06  2006-07-03
      Added OTU's to Tissue, and setting of gaslist

0.05  2006-06-29
      Made test case more robust, changed nodeco_time handling

0.03  2006-06-26
      Finished POD. Checked Makefile.PL and cleaned up email addresses.

0.01  Thu May  4 17:22:07 2006
	- original version; created by h2xs 1.23 with options
		-A -c -n Deco




Makefile.PL
use 5.008004;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Deco',
    VERSION_FROM      => 'lib/Deco.pm', # finds $VERSION
    PREREQ_PM         => {
		Config::General   => 0,
		GD::Graph         => 0,
		Test::Exception   => 0,
		Test::More        => 0,
		POSIX	          => 0,
    },
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Deco.pm', # retrieve abstract from module
       AUTHOR         => 'Jaap Voets <narked@xperience-automatisering.nl>') : ()),
    EXE_FILES  => [
        'bin/diveplot.pl',
		   ],
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);



lib/Deco.pm
package Deco;

use strict;
use warnings;

our $VERSION = '0.11';

1;
__END__

=head1 NAME

Deco - Module for simulating body tissue during a scuba dive

=head1 SYNOPSIS

  use Deco::Dive
  my $dive = Deco::Dive->new( model => 'haldane' );
  $dive->load_data_from_file( file => 'profile.txt' );
  $dive->gas( 'O2' => 40, 'N2' => 60);	
  $dive->simulate();

=head1 DESCRIPTION

The Deco package itself does not do anything useful, it only serves as the root of the other packages. You will want to look into the Deco::Dive module as that is the one to be using directly in your own scripts.

=head2 EXPORT

None by default.

=head1 SEE ALSO

An extensive treatment of the Haldane decompression theory is prodived in the Deco.pdf document which can be found in the docs directory.

=head1 AUTHOR

Jaap Voets, E<lt>narked@xperience-automatisering.nlE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Jaap Voets

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Deco/Dive.pm
#######################################
# Module  : Deco::Dive.pm
# Author  : Jaap Voets
# Date    : 27-05-2006
# $Revision$
#######################################
package Deco::Dive;

use strict;
use warnings;
use Carp;
use Config::General;
use Deco::Tissue;

our $VERSION = '0.3';

our @MODELS = ('haldane', 'padi', 'usnavy');

# Constructor
sub new {
    my $class = shift;
    my %args  = @_;

    my $self = {};

    # the data points for the dive, both arrays
    $self->{timepoints} = [];
    $self->{depths}     = [];

    # an array of tissues to use
    $self->{tissues}    = ();
    
    # super structure to remember all tissue info per timepoint
    $self->{info}       = {};

    # where can we find the config?
    $self->{config_dir} = $args{configdir} || '.';

    # theoretical tissue model we'll be using
    $self->{model}        = '';
    $self->{model_name}   = '';
    bless $self, $class;
    
    return $self;
}

# load the dive profile data from a file
sub load_data_from_file {
    my $self = shift;
    my %opt  = @_;

    my $file = $opt{file};
    croak "No file specified, to load dive profile" unless $file;
    # check whether the file exists
    croak "File $file does not exist" unless ( -e $file);

    # field separator
    my $sep        = $opt{separator} || ';';
    my $timefield  = $opt{timefield} || '0';
    my $depthfield = $opt{depthfield} || 1;
    my $timefactor = $opt{timefactor} || 1; # factor to get each time point in seconds

    my (@times, @depths);
    open (IN, $file) || croak "Can't open file $file for reading";
    while (my $line = <IN>) {
	chomp($line);
	next if $line =~ /^\s*#/; # skip comment lines
	next if $line =~ /^\s+$/; # skip empty lines
	my @fields = split(/$sep/, $line);
	push @times, $timefactor * $fields[$timefield];
	my $depth = $fields[$depthfield];
	if ($depth < 0) {
	    $depth = -1 * $depth;
	}
	push @depths, $depth;
    }
    close(IN);
    
    $self->{depths}     = \@depths;
    $self->{timepoints} = \@times;
    
}

# set a time, depth point
sub point {
    my $self = shift;
    my ($time, $depth) = @_;
    push @{ $self->{depths} }, $depth;
    push @{ $self->{timepoints} }, $time;
}

# pick a model and load the corresponding config
# this will create a list of tissues
# either specify a config file and read the model from there
#  - or - specify a model and read in the default file
sub model {
    my $self = shift;
    my %opt  = @_;

    my ($config_file, $model);
    if ( $opt{config} ) {
	$config_file = $opt{config};
	# model will be read from config
    } elsif ( $opt{model} ) {
	$model = lc( $opt{model} );
	$config_file = $self->{config_dir} . "/$model.cnf";
    } else {
	croak "Please specify the config file or model to use!";
    }

    # load the config
    my $conf   = new Config::General(  -ConfigFile => $config_file,  -LowerCaseNames => 1 );
    my %config = $conf->getall;
 
    $model = lc($config{model});

    # remember the model we use
    $self->{model}      = $model;
    $self->{model_name} = $config{name};

    croak "Invalid model $model" unless grep { $_ eq $model } @MODELS;
    
    # cleanup first
    $self->{tissues} = ();

    # create all the tissues
    foreach my $num (keys %{ $config{tissue} }) {
	$self->{tissues}[$num] = new Deco::Tissue( halftime => $config{tissue}{$num}{halftime}, 
						   M0       => $config{tissue}{$num}{m0}, 
						   deltaM   => $config{tissue}{$num}{deltam} ,
						   nr       => $num,
						   );
    }
    
    return 1;
}

# run the simulation
sub simulate {
    my $self = shift;
    my %opt  = @_;

    # model passed to us takes precedence, if that is not present
    # we see if the model was already set, otherwise we default to haldane
    my $model = lc($opt{model}) || $self->{model} || 'haldane';
    croak "Invalid model $model" unless grep { $_ eq $model } @MODELS;
    
    # first load the model
    $self->model( model => $model,  config => $self->{config_dir} . '/' . $model . '.cnf');
    
    # then check whether we loaded data
    if ( scalar( @{ $self->{timepoints} } ) == 0 ) {
	croak "No dive profile data present, forgot to call dive->load_data_from_file() ?";
    }
    
    # step through all the timepoints & depths
    my $i = 0;
    my @times  = @{ $self->{timepoints} };
    my @depths = @{ $self->{depths} };
    foreach my $time ( @times ) {
	# get the corresponding depth
	my $depth = $depths[$i];
	$i++;

	my $nodeco_dive          = 1000000;
	my $leading_tissue_deco  = '';

	my $safe_depth_dive      = 0;
	my $leading_tissue_depth = '';

	# loop over all the tissues
	foreach my $tissue ( @{ $self->{tissues} } ) {
	    next if ! defined $tissue;

	    my $num  = $tissue->nr;

	    $tissue->point( $time, $depth );
	    
	    # we like to have 
	    # no_deco time, is special, it can return - for not applicable
	    my $nodeco = $tissue->nodeco_time();
	    $nodeco = undef if $nodeco eq '-';
	    $self->{info}->{$num}->{$time}->{nodeco_time}    = $nodeco; 
	    if ($nodeco) {
		if ($nodeco < $nodeco_dive) {
		    $nodeco_dive         = $nodeco;
		    $leading_tissue_deco = $tissue->nr();
		}
	    }

	    # safe depth, meters, positive
	    my $safe_depth = $tissue->safe_depth();
	    $self->{info}->{$num}->{$time}->{safe_depth} =  $safe_depth;
	    if ($safe_depth > $safe_depth_dive) {
		$safe_depth_dive      = $safe_depth;
		$leading_tissue_depth = $tissue->nr();
	    }

	    # percentage filled compared to M0 pressure
	    $self->{info}->{$num}->{$time}->{percentage} =  $tissue->percentage();

	    # internal pressure
	    $self->{info}->{$num}->{$time}->{pressure}   = $tissue->internalpressure();

	    # OTU's
	    $self->{info}->{$num}->{$time}->{pressure}   = $tissue->calculate_otu();

	}
	if ($nodeco_dive == 1000000) {
	    $nodeco_dive = '-';
	}
	$self->{info}->{dive}->{$time}->{nodeco}             = $nodeco_dive;
	$self->{info}->{dive}->{$time}->{leadingtissuedeco}  = $leading_tissue_deco;
	$self->{info}->{dive}->{$time}->{safedepth}          = $safe_depth_dive;
	$self->{info}->{dive}->{$time}->{leadingtissuedepth} = $leading_tissue_depth;

    }
    
}

# set gas fractions
sub gas {
    my $self = shift;
    my %gaslist = @_;
    # just pass it off to each tissue
    foreach my $tissue ( @{ $self->{tissues} } ) {
        next if ! defined $tissue;

	# the tissue module will croak on setting wrong gas
	# just let it bubble up to the calling script from here
	$tissue->gas( %gaslist );
    }
}

# calculate the no-deco time for the dive
# this will be the smalles value of the nodeco times of
# the tissues of this model
#
# time is minutes, it takes the current depth and time of the tissue
# second return value is the tissue nr that gave the minimal nodeco_time
sub nodeco_time {
    my $self = shift;
    # loop over all the tissues
    my $nodeco_time = 1000000; # start with absurd high value for easy comparing
    my $tissue_nr   = '';
    foreach my $tissue ( @{ $self->{tissues} } ) {
	next if ! defined $tissue;
	my $time = $tissue->nodeco_time();
	if ($time ne '-') {
	    if ($time < $nodeco_time) {
		$nodeco_time = $time;
		$tissue_nr   = $tissue->nr();
	    }
	}
    }
    if ($nodeco_time == 1000000) {
	$nodeco_time = '-';
    }
    return ($nodeco_time, $tissue_nr);
}

1;


__END__

=head1 NAME

Dive - Simulate a dive and corresponding tissues

=head1 SYNOPSIS

    use Deco::Dive;
my $dive = new Deco::Dive( );
$dive->load_data_from_file( file => $file);
$dive->model( config => '/path/to/my/model.cnf' );
$dive->simulate( );


=head1 DESCRIPTION

The Dive model can be used to simulate a dive. You add data points, set some properties and call the simulate method to calculate the entire dive.

After simulating, you can retrieve info in several ways from the dive.



=head2 METHODS

=over 4

=item new()

The constructor of the class.

=item $dive->load_data_from_file( file => $file , timefield => 0, depthfield => 1, timefactor => 1, separator => ';');

Load data from a csv file. You HAVE to specify the filename. Additional options are timefield, the 0 based field number where the  timestamps are stored. Depthfield, field number where the depth (in meters is stored), separator, the fieldseparator and timefactor, the factor to multiply the time field with to transform them to seconds.

=item $dive->model( model => 'padi', config => $file );

Set the model to use. If you specify one of the known models and the config dir has been set right,
then the method will load the corresponding config file and set up the tissues for this model.

Alternatively you can specify your own config file to use.

=item $dive->simulate( model => 'haldane' );

This method does the simulation for all tissues for the chosen model. It will run along all the time and depth
points of the dive and calculate gas loading for all the tissues of the model.

=item $dive->nodeco_time();
This function will loop over all the tissues of the model, calling the LDeco::Tissue::nodeco_time() function on them. The lowest value will be stored, together with the associated tissue nr.  

=item $dive->gas( 'O2' => 45, 'n2' => 0.55);

Set the gases used during this dive. Currently supported are 02, N2 and He. Enter the fraction of the gas either as real fraction, or as a percentage.

=item $dive->point($time, $depth);

Adds a new point to the dive. Use it if you don't want to load data from a file but iterate over your own values. Time should be in seconds, depth in meters.
 
=back

=head2 EXPORT

None by default.

=head1 SEE ALSO

L<Deco>, L<Deco::Tissue>, L<Deco::Dive::Plot>. L<SCUBA::Table::NoDeco> might be of interest to you as well.
In the docs directory you will find an extensive treatment of decompression theory in the file Deco.pdf. A lot of it has been copied from the www.deepocean.net website.

=head1 AUTHOR

Jaap Voets, E<lt>narked@xperience-automatisering.nlE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Jaap Voets

=cut


lib/Deco/Dive/Table.pm
#######################################
# Module  : Deco::Dive::Table.pm
# Author  : Jaap Voets
# Date    : 10-10-2006
#######################################
package Deco::Dive::Table;

use strict;
use warnings;
use Carp;

use constant INITIAL_NODECO => 1000000;
our $VERSION = '0.1';


# Constructor
sub new {
    my $class = shift;
    my $dive = shift;   

    croak "Please provide a Deco::Dive object for plotting" unless ref($dive) eq 'Deco::Dive';

    my $self = { dive => $dive };
		
    # depth in meters, we will use these for the table
    # can be overridden by ->setdepths
    $self->{depths} = [10, 12, 14, 16, 18, 20, 24, 27, 30, 33, 36, 40, 42, 45, 50];
    
    # place to store our numbers
    $self->{table} = undef;
    bless $self, $class;
    
    return $self;
}

# set the depths for the table you want
sub setdepths {
    my $self = shift;
    $self->{depths} = \@_;
}

sub calculate {
    my $self = shift;
    
    foreach my $depth ( @{ $self->{depths} } ) {
	my $nodeco_min = INITIAL_NODECO;
	foreach my $tissue ( @{ $self->{dive}->{tissues} } ) {
	    next if ! defined $tissue;   # first array element is empty
	    
	    # we go instantly to the depth and ask for the no_deco time
	    $tissue->point( 0, $depth );	
	    
	    # we like to have 
		    # no_deco time, is special, it can return - for not applicable
	    my $nodeco = $tissue->nodeco_time();
	    $nodeco = undef if $nodeco eq '-';
	    
	    if ($nodeco) {
		if ($nodeco < $nodeco_min) {
		    $nodeco_min = int($nodeco);	
		}	
	    } 
    	}
    	if ($nodeco_min == INITIAL_NODECO) {
	    $nodeco_min = '-';
    	}
	
	$self->{table}->{$depth} = $nodeco_min; 
    }
    
}

sub output {
	my $self = shift;
	my %opt  = @_;
	
	my $template = $opt{template} || 'No Decompression limit at #DEPTH# is #TIME# minutes';
	my $output = '';
	foreach my $depth ( @{ $self->{depths} }) {
		my $row = $template;
		$row =~ s/#DEPTH#/$depth/gi;
		$row =~ s/#TIME#/$self->{table}->{$depth}/gi;
		$output .= $row;
	}	
	
	return $output;
}
1;


__END__

=head1 NAME

Deco::Dive::Table - Generate a list of no stop limits for your model 

=head1 SYNOPSIS

    use Deco::Dive;
    use Deco::Dive::Table;
    
    my $dive = new Deco::Dive( );
    $dive->model( config => './conf/haldane.cnf');

    my $divetable = new Deco::Dive::Table( dive => $dive );
    $divetable->calculate();
    my $table = $divetable->output();

=head1 DESCRIPTION

This package will plot the profile of the dive and internal pressures of the tissues of the model.


=head2 METHODS

=over 4

=item $divetable->new( dive => $dive );

The constructor of the class. There is only one parameter: a Deco::Dive object.

=item $divetable->setdepths( $depth1, $depth2, $depth3, ....  );

Set the list of depths you want the table to be for manually. There is a default list provided, but with this method you can overrule it. Depths should be entered in B<meters>

=item $divetable->calculate();

Performs the calculation of the table. You will need to call this function before retrieving output.

=item $divetable->output( [template => $template ] );

Retrieve the table as a string. Optionally you can supply your own template for each line of output.
The placeholders #DEPTH# and #TIME# will be replaced by the actual values for the depth (in meters) and time (in minutes) that you can stay at that depth without required decompression stops. 

=back

=head2 EXPORT

None by default.

=head1 SEE ALSO

In the docs directory you will find an extensive treatment of decompression theory in the file Deco.pdf. A lot of it has been copied from the www.deepocean.net website.

=head1 AUTHOR

Jaap Voets, E<lt>narked@xperience-automatisering.nlE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Jaap Voets

=cut


lib/Deco/Dive/Plot.pm
#######################################
# Module  : Deco::Dive::Plot.pm
# Author  : Jaap Voets
# Date    : 02-06-2006
#######################################
package Deco::Dive::Plot;

use strict;
use warnings;
use Carp;
use GD::Graph::lines;
use GD::Graph::bars;

our $VERSION = '0.3';

# some constants used 
use constant DEFAULT_WIDTH  => 600;
use constant DEFAULT_HEIGHT => 400;

# Constructor
sub new {
    my $class = shift;
    my $dive = shift;   

    croak "Please provide a Deco::Dive object for plotting" unless ref($dive) eq 'Deco::Dive';

    my $self = { dive => $dive };
    
    bless $self, $class;
    
    return $self;
}

# plot the depth versus time 
sub depth {
    my $self = shift;
    my %opt  = @_;
    
    # divide the seconds by 60 to get minutes
    my @times = map { $_ / 60 } @{ $self->{dive}->{timepoints} };
    croak "There are no timestamps set for this dive" if ( scalar( @times ) == 0);

    # multiply the depths by -1 to get a nicer picture
    my @depths  = map { -1 * $_ } @{ $self->{dive}->{depths} };
    croak "There are no depth points set for this dive" if ( scalar( @depths ) == 0);

    my $width  = $opt{width}  || DEFAULT_WIDTH;
    my $height = $opt{height} || DEFAULT_HEIGHT;
    my $outfile = $opt{file}  || 'depth.png';

    my $graph =  GD::Graph::lines->new($width, $height);
    $graph->set(
             x_label           => 'Time (minutes)',
             y_label           => 'Depth (meter)',
             title             => 'Depth profile',
	     y_max_value       => 0,	
	    ) or die $graph->error;

    my @data = (\@times, \@depths);

    my $gd = $graph->plot(\@data) or die $graph->error;
    open(IMG, ">$outfile") or die $!;
    binmode IMG;
    print IMG $gd->png;
    close IMG;

}

# plot the percentage saturation of each tissue
# this will be a bar graph using the tissue nr for the x-axis
sub percentage {
	my $self = shift;	
	my %opt  = @_;
	
	# at what point of the dive do we
	my $plot_time;
	if (! defined $opt{time} ) {
		# take the last one
		$plot_time = pop @{ $self->{dive}->{timepoints} };
	} else {
	    # find the corresponding time in seconds in the time array
		my $time = 60 * $opt{time}; # time in minutes
		foreach my $timestamp ( @{ $self->{dive}->{timepoints} } ) {
			if ($timestamp <= $time ) {
				$plot_time = $timestamp;
			}	
		}	
	}
	 
	# get the data
	my @nrs;
	my @percentages;
	foreach my $tissue ( @{ $self->{dive}->{tissues} } ) {
		next if ! defined $tissue;
		my $num  = $tissue->nr;
		push @nrs, $num;	
		# peek inside the Dive object to get the precalculated percentages  
		push @percentages, sprintf('%.0f', $self->{dive}->{info}->{$num}->{$plot_time}->{percentage});
	}
    croak "There is nothing to display for this dive" if ( scalar( @percentages ) == 0);
    
    my $width  = $opt{width}  || DEFAULT_WIDTH;
    my $height = $opt{height} || DEFAULT_HEIGHT;
    my $outfile = $opt{file}  || 'percentage.png';
	
	my $graph =  GD::Graph::bars->new($width, $height);
    $graph->set(
             x_label           => 'Tissue',
             y_label           => 'Percentage',
             title             => 'Tissue saturation',
	     	 y_min_value       => 0,	
	    ) or die $graph->error;

    my @data = (\@nrs, \@percentages);
    my $gd = $graph->plot(\@data) or die $graph->error;
    open(IMG, ">$outfile") or die $!;
    binmode IMG;
    print IMG $gd->png;
    close IMG;
}


sub pressures {
    my $self = shift;
    my %opt  = @_;

    $opt{y_label} = 'Internal Pressure (bar)';
    $self->_info( 'pressure',  %opt );
}

sub nodeco {
    my $self = shift;
    my %opt  = @_;

    $opt{y_label} = 'No deco time (minutes)';
    $self->_info( 'nodeco_time',  %opt );
}

# plot a certain info series for all tissues
# after simulating a dive, there are arrays of information setup
# throught this routine you can get the series of each info
sub _info {
    my $self = shift;
    my $what = shift; # one of nodeco_time, safe_depth, percentage or pressure
    my %opt  = @_;

    my @times = @{ $self->{dive}->{timepoints} };
    croak "There are no timestamps set for this dive" if ( scalar( @times ) == 0);
    
    # divide the seconds by 60 to get minutes
    my @minutes = map { $_ / 60 } @{ $self->{dive}->{timepoints} };
    
    my $width  = $opt{width}  || DEFAULT_WIDTH;
    my $height = $opt{height} || DEFAULT_HEIGHT;
    my $outfile = $opt{file}  || $what . '.png';
    
    my $y_label = $opt{y_label} || 'Depth (meter)';
    my $graph =  GD::Graph::lines->new($width, $height);
    $graph->set(
		x_label           => 'Time (minutes)',
		y_label           => $y_label,
		title             => "$what profile",
		) or die $graph->error;
    
    my @data;
    push @data, \@minutes; # load the time values

    foreach my $tissue ( @{ $self->{dive}->{tissues} } ) {
	next if ! defined $tissue;   # first array element is empty
	my $num = $tissue->nr;

	my @y = ();
	foreach my $time (@times) {
	    push @y, $self->{dive}->{info}->{$num}->{$time}->{$what};
	}

	# add the series to the plot data
	push @data, \@y;
    }
    
    my $gd = $graph->plot(\@data) or die $graph->error;
    open(IMG, ">$outfile") or die $!;
    binmode IMG;
    print IMG $gd->png;
    close IMG;

}

1;


__END__

=head1 NAME

Dive - Simulate a dive and corresponding tissues

=head1 SYNOPSIS

    use Deco::Dive;
    use Deco::Dive::Plot;
    
    my $dive = new Deco::Dive( );
    $dive->load_data_from_file( file => $file);
    $dive->simulate( model => 'haldane');

    my $diveplot = new Deco::Dive::Plot( dive => $dive );
    $diveplot->depth( file => 'depth.png' );
    $diveplot->pressures( file => 'pressures.png' );

=head1 DESCRIPTION

This package will plot the profile of the dive and internal pressures of the tissues of the model.


=head2 METHODS

=over 4

=item new( dive => $dive )

The constructor of the class. Takes a required parameter: a Deco::Dive object.

=item $diveplot->depth( width=> $width, height => $height, file => $file );

Plots the depth versus time graph of the dive. It will default to a file called depth.png in 
the current directory, with a size of 600 x 400 pixels.

=item $diveplot->pressures( width=> $width, height => $height, file => $file );

This method will plot the internal pressures of all the tissues of the model during the dive.

=item $diveplot->percentage( time => $time, width=> $width, height => $height, file => $file );

This function will plot a bar graph of the pertenage of saturation (o fthe allowed value per tissue).
You can specify the time in minutes for the point in the dive where you want to see the percentages for.

=item $diveplot->nodeco( width=> $width, height => $height, file => $file );

This method will plot the no deco time during the dive for each tissue

=back

=head2 EXPORT

None by default.

=head1 SEE ALSO

L<Deco>, L<Deco::Tissue>, L<Deco::Dive>. L<SCUBA::Table::NoDeco> might be of interest to you as well.

In the docs directory you will find an extensive treatment of decompression theory in the file Deco.pdf. A lot of it has been copied from the www.deepocean.net website.

=head1 AUTHOR

Jaap Voets, E<lt>narked@xperience-automatisering.nlE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Jaap Voets

=cut
